### ISSUE TITLE
 Specifying modules in compilers.yaml doesn't preserve ENV variables

### ISSUE THREAD

**Issue by pramodk** [spack/spack#2056]
This was discussed/mentioned in the #2037. In short, when I specify module in `compilers.yaml` and if that module is setting some ENV variables like `LD_LIBRARY_PATH` then those are not preserved.

Summary:

pramodk:

> if I specify modules in compilers.yaml, does all env variables from those modules will be set and preserved during build? Or this also requires --dirty ?

tgamblin:

> @pramodk: if you specify modules in the compilers.yaml file, they should be imported into the build environment even without --dirty. Spack has (as of the Cray support) the ability to load modules just in the build environment.
> @mamelara @becker33: do you know why that might not be working? This is on a Linux machine -- did you test there?

pramodk:

> In my testing it's not working as expected. Here is detailed example (x86_redhat_cluster):
> 
> I have `gcc` module which looks like this:
> 
> ``` bash
> module show gcc/4.9.0
> -------------------------------------------------------------------
> /home/modulefiles/gcc/4.9.0:
> 
> 
> 
> module-whatis  Loads gcc 4.9.0 to your environment variables
> module         load mpc/1.0.2
> module         load mpfr/3.1.2
> module         load gmp/5.1.3
> prepend-path   PATH /home//tools/gcc/gcc-4.9.0/install/bin
> prepend-path   LD_LIBRARY_PATH /home//tools/gcc/gcc-4.9.0/install/lib64
> prepend-path   LD_RUN_PATH /home//tools/gcc/gcc-4.9.0/install/lib64
> prepend-path   CPATH /home//tools/gcc/gcc-4.9.0/install/include
> prepend-path   SEARCH_DIR /home//tools/gcc/gcc-4.9.0/install/lib64
> ```
> 
> If I use that above compiler directly (without loading module), I get following error:
> 
> ```
> /home//tools/gcc/gcc-4.9.0/install/bin/gcc example.c
> /home/tools/gcc/gcc-4.9.0/install/bin/../libexec/gcc/x86_64-unknown-linux-gnu/4.9.0/cc1: error while loading shared libraries: libmpc.so.3: cannot open shared object file: No such file or directory
> ```
> 
> So `LD_LIBRARY_PATH` is not set for `mpc` and hence the error which is expected.
> 
> Now I add entry into `compilers.yaml` as:
> 
> ``` bash
> - compiler:
>     modules: [gcc/4.9.0, mpc/1.0.2, mpfr/3.1.2, gmp/5.1.3]
>     operating_system: redhat6
>     paths:
>       cc: /home/tools/gcc/gcc-4.9.0/install/bin/gcc
>       cxx: /home/tools/gcc/gcc-4.9.0/install/bin/g++
>       f77: /home/tools/gcc/gcc-4.9.0/install/bin/gfortran
>       fc: /home/tools/gcc/gcc-4.9.0/install/bin/gfortran
>     spec: gcc@4.9.0
> ```
> 
> (I also tried adding only gcc/4.9.0).
> 
> If I try to build a package, I see following error in `pack-build.out`:
> 
> ``` bash
> $ spack install -v  testpkg %gcc@4.9.0
>   /home/tools/gcc/gcc-4.9.0/install/bin/../libexec/gcc/x86_64-unknown-linux-gnu/4.9.0/cc1:
>   error while loading shared libraries: libmpc.so.3: cannot open shared
>   object file: No such file or directory
> ```
> 
> I haven't loaded any module or set any path yet. But if I just run with `--dirty` flag:
> 
> ``` bash
> spack install -v --dirty testpkg %gcc@4.9.0
> ```
> 
> Now the package build successfully! My understanding was that the `--dirty` flag will be only useful if I load any modules or set any env variables.
> 
> If I check `spack-build.env` in above two cases, `LD_LIBRARY_PATH` is set properly / preserved if I use `--dirty` flag (even without loading any module or setting path explicitly).

**Comment by citibeth** [comment]
Yes, you have to use the `--dirty` flag.  I've experienced a similar issue on one of our systems.

**Comment by davydden** [comment]
looks like `--dirty` does not always help: https://github.com/LLNL/spack/issues/2037#issuecomment-254561822

**Comment by pramodk** [comment]
@citibeth @davydden : Isn't `module` supposed to load and preserve environment variables? If so, I will keep this issue open as bug. From my understanding this is how it works for Cray system (if I haven't misunderstood anything).

**Comment by davydden** [comment]
i think @adamjstewart suggested that the problem is due to

> The compiler wrappers themselves delete LD_LIBRARY_PATH during the final stage. That's probably the source of the problem.

**Comment by davydden** [comment]
here's the code where those variables are unset https://github.com/LLNL/spack/blob/develop/lib/spack/env/cc#L323-L325

**Comment by pramodk** [comment]
thanks @davydden! I looked at the wrappers but one thing that is not clear to me when `modules` come into effect i.e. `modules` are loaded and then `compiler` wrappers clean those ENV variables? 
If so how that works on Cray system? Because on Cray machines PGI compiler looks like:

```bash
$ module show pgi/16.3.0
-------------------------------------------------------------------
/opt/modulefiles/pgi/16.3.0:
setenv		 PGI_PATH /opt/pgi/16.3.0
setenv		 PGI /opt/pgi/16.3.0
prepend-path	 PGROUPD_LICENSE_FILE /opt/pgi/license.dat
prepend-path	 PATH /opt/pgi/16.3.0/linux86-64/16.3/bin
prepend-path	 MANPATH /opt/pgi/16.3.0/linux86-64/16.3/man
prepend-path	 LD_LIBRARY_PATH /opt/pgi/16.3.0/linux86-64/16.3/lib
```
And if I remember correctly on Cray this is working fine.

**Comment by davydden** [comment]
> modules are loaded and then compiler wrappers clean those ENV variables? 

i have no idea ... ;-( Would like to know myself as well...

**Comment by pramodk** [comment]
@tgamblin , @adamjstewart , @alalazo : help to understand this better?

**Comment by pramodk** [comment]
As discussed in #2072, with #2074 `--dirty` works in all cases (at least what I am testing with PGI, GCC modules which set LD_LIBRARY_PATH)

**Comment by davydden** [comment]
@pramodk out of curiosity, why do you still need `--dirty` if you specify `modules: [gcc/4.9.0, mpc/1.0.2, mpfr/3.1.2, gmp/5.1.3]` in `compilers.yaml`? I think now everything is supposed to work without `--dirty`?

p.s. shall this issue be closed?

**Comment by pramodk** [comment]
@davydden : that still confuses me! I have : 
```bash
- compiler:
    modules: [gcc/4.9.0, mpc/1.0.2, mpfr/3.1.2, gmp/5.1.3]
    operating_system: redhat6
    paths:
      cc: /home/kumbhar/tools/gcc/gcc-4.9.0/install/bin/gcc
      cxx: /home/kumbhar/tools/gcc/gcc-4.9.0/install/bin/g++
      f77: /home/kumbhar/tools/gcc/gcc-4.9.0/install/bin/gfortran
      fc: /home/kumbhar/tools/gcc/gcc-4.9.0/install/bin/gfortran
    spec: gcc@4.9.0
```
And `spack install` gives : 
```bash
  error while loading shared libraries: libmpc.so.3: cannot open shared
  object file: No such file or directory
```
If I look at `spack-build.env`, there is no `LD_LIBRARY_PATH` set.

I haven't loaded `gcc/4.9.0` and other modules in `bash`. But, just adding `--dirty` solves this issue and I see correct `LD_LIBRARY_PATH` in `spack-build.env`.

@tgamblin : Is #2074 supposed to solve the issue with module loaded compilers? For above tests, I still needed to add `--dirty` flag.

**Comment by davydden** [comment]
@pramodk I think something is still wrong. If you specify loading of modules in `compilers.yaml`, then they should be loaded and available during the build in Spack. 

Here is a [quote](https://github.com/LLNL/spack/issues/2037#issuecomment-254346248) from @tgamblin : 

> if you specify modules in the compilers.yaml file, they should be imported into the build environment even without --dirty. Spack has (as of the Cray support) the ability to load modules just in the build environment.

@mamelara @becker33: could you please comment on this issue?

**Comment by tgamblin** [comment]
@pramodk: is this fixed by #2072?

**Comment by citibeth** [comment]
Would it be reasonable to write these wrappers in Python, like the rest of
Spack?

**Comment by tgamblin** [comment]
@citibeth: the compiler wrappers?

**Comment by alalazo** [comment]
@citibeth If you refer to the compiler wrappers : I guess using python instead of a shell script will be a major slowdown given the number of times they are called during a build...

**Comment by tgamblin** [comment]
@citibeth @alalazo: fun fact: the compiler wrappers were originally in Python.  See 0f3b80cddbfd775c10f68423444aca792525f856.  Python startup overhead is really expensive, especially if your python install is on a shared filesystem.  As I recall making this change sped the builds up by as much as 6x.

**Comment by citibeth** [comment]
Thanks... yes, I withdraw my suggestion.  I don't want builds to be 6x
slower.

**Comment by pramodk** [comment]
@tgamblin : 
```
@pramodk: is this fixed by #2072?
```
No, I tested this again and still see the issue! Below are the steps:

* In bash I have following env var set:
```
echo $LD_LIBRARY_PATH
/home/kumbhar/softwares/install/lib
```
* the `mpi` and compiler modules specified in `packages.yaml` are setting following env variable : 
```
LD_LIBRARY_PATH /tools/pgi/16.5/linux86-64/2016/mpi/openmpi-1.10.2/lib
LD_LIBRARY_PATH /tools/pgi/16.5/linux86-64/16.5/lib
```

* If I look at `spack-build.env` file generated by `spack install` with and without `--dirty` flag, the diff is : 
```
diff /tmp/kumbhar/spack-stage/spack-stage-R5zpjB/mypkg/spack-build.env.no-dirty /tmp/kumbhar/spack-stage/spack-stage-R5zpjB/mypkg/spack-build.env
30a31
> LD_LIBRARY_PATH=/tools/pgi/16.5/linux86-64/2016/mpi/openmpi-1.10.2/lib:/tools/pgi/16.5/linux86-64/16.5/lib:/home/kumbhar/softwares/install/lib
```

I see `unset('LD_LIBRARY_PATH')` below. If remove that (which then become equivalent of `--dirty` I assume), LD_LIBRARY_PATH is maintained:
```bash
--- a/lib/spack/spack/build_environment.py
+++ b/lib/spack/spack/build_environment.py
@@ -284,7 +284,7 @@ def set_build_environment_variables(pkg, env, dirty=False):
         # Remove these vars from the environment during build because they
         # can affect how some packages find libraries.  We want to make
         # sure that builds never pull in unintended external dependencies.
-        env.unset('LD_LIBRARY_PATH')
+        #env.unset('LD_LIBRARY_PATH')
```

**Comment by alalazo** [comment]
Environment should now be cleaned before loading modules for compilers:
https://github.com/spack/spack/blob/745c191d7362b221875412e640e02db68dedc12d/lib/spack/spack/build_environment.py#L791-L836
